parsers/verilog.c

```diff
@@ -420,25 +420,49 @@ static short hasSimplePortList (verilogKind kind)
 	}
 }
 
-static tokenInfo *newToken (void)
+static void clearToken (tokenInfo *token)
 {
-	tokenInfo *const token = xMalloc (1, tokenInfo);
-	token->kind = K_UNDEFINED;
-	token->name = vStringNew ();
+	token->kind = K_UNDEFINED;	// to be set by updateKind()
+	vStringClear (token->name);
 	token->lineNumber = getInputLineNumber ();
 	token->filePosition = getInputFilePosition ();
 	token->scope = NULL;
 	token->nestLevel = 0;
 	token->lastKind = K_UNDEFINED;
-	token->blockName = vStringNew ();
-	token->inheritance = vStringNew ();
+	vStringClear (token->blockName);
+	vStringClear (token->inheritance);
 	token->prototype = false;
 	token->classScope = false;
 	token->parameter = false;
 	token->hasParamList = false;
+}
+
+static tokenInfo *newToken (void)
+{
+	tokenInfo *const token = xMalloc (1, tokenInfo);
+	token->name = vStringNew ();
+	token->blockName = vStringNew ();
+	token->inheritance = vStringNew ();
+	clearToken(token);
 	return token;
 }
 
+static tokenInfo *dupToken (tokenInfo *token)
+{
+	tokenInfo *dup = newToken ();
+	vStringDelete (dup->name);
+	vStringDelete (dup->blockName);
+	vStringDelete (dup->inheritance);
+	*dup = *token;
+	dup->name = vStringNew ();
+	dup->blockName = vStringNew ();
+	dup->inheritance = vStringNew ();
+	vStringCopy (dup->name, token->name);
+	vStringCopy (dup->blockName, token->blockName);
+	vStringCopy (dup->inheritance, token->inheritance);
```

Comment:
You does "delete", "new", and "copy" three vStrings.

However, when you call newToken () at the beginning of the funciton,  the `dup` returned from newToken() is already cleared.
So the "delete" and "new" steps are just redundant. What you need is only the the "copy" step.

Reply:
Thank you for your review.

First I did make same mistake as you wrote.  But it does not work.
`dup->name`, `dup->blockName`, and `dup->inheritance` are overwritten by `token->name`, `token->blockName`, and `token->inheritance` via `*dup = *token;`.
vStrings allocated by `dupToken()` are never be freed.

Another solution is copying each element without using  `*dup = *token;`.  But I did not choose it because it might cause missing elements.  I am assuming the number of vString elements are much less than the number of other elements.

Reply:
How about the following code?

```
static tokenInfo *dupToken (tokenInfo *token)
{
	tokenInfo *dup = newToken ();

	{
	  tokenInfo tmp = *dup;

	  *dup = *token;

	  dup->name = tmp.name;
	  dup->blockName = tmp.blockName;
	  dup->inheritance = tmp.inheritance;
	}

	vStringCopy (dup->name, token->name);
	vStringCopy (dup->blockName, token->blockName);
	vStringCopy (dup->inheritance, token->inheritance);
	return dup;
}
```

Reply:
Thanks.  Better than mine.
I've committed the following code.

``` C
static tokenInfo *dupToken (tokenInfo *token)
{
	tokenInfo *dup = newToken ();
	tokenInfo tmp = *dup;	// save vStrings, name, blockName, and inheritance
	*dup = *token;
	// revert vStrings allocated for dup
	dup->name = tmp.name;
	dup->blockName = tmp.blockName;
	dup->inheritance = tmp.inheritance;
	// copy contents of vStrings
	vStringCopy (dup->name, token->name);
	vStringCopy (dup->blockName, token->blockName);
	vStringCopy (dup->inheritance, token->inheritance);
	return dup;
}
```